[Seven] Conventions docs (#7375)
plone/volto: docs/reference/conventions.md

    "property=og:description": "Development conventions, rules and best practices"
    "property=og:title": "Development conventions, rules and best practices"
    "keywords": "Plone, Seven, frontend, best practices, conventions, rules"
---

# Development conventions, rules and best practices

This document explains the conventions, rules, and best practices we follow in the Plone 7 frontend codebase.
We generally stick to the same models, artifacts, and conventions as in Volto, unless there's a good reason to change.
For example, the blocks model structure stays the same since it's well established and widely used.
Another example: in Volto, we use controlled forms everywhere.
Even though React Router 7 introduced a new way to handle forms, we stick to controlled forms for consistency.
Our widgets also follow almost the same interface as in Volto because controlled forms need controlled widgets.
The configuration system remains based on `@plone/registry`, and we keep improving it.
We follow the mantra: "if it works, don't change it."
This document highlights the differences and additions compared to Volto conventions.

> Separate builds for public and CMSUI

Reviewer: Is this supposed to be a heading or a sentence? If a heading, it needs markup and content.

Author: It was supposed that I elaborate that idea. I will.

This document highlights the differences and additions compared to Volto conventions.

> ## Separate builds for public and CMSUI

Reviewer: I know, it's nitpicky, but it helps make the distinction.

```suggestion
## Separate builds for public UI and CMS UI
```

Author: Ok for the PUBLIC UI but, CMSUI is something that the community coined and I'd like to maintain it like this.

> We are aiming for the possibility of having two separate builds: one for the public theme and another for the CMS UI.

Reviewer: ```suggestion
We aim for the possibility of having two separate builds: one for the public UI as a theme and another for the CMS UI.
```

Thus, we need to ensure that the codebase can support this separation and have this in mind when developing new features.
This will impose some constraints, such as avoiding the use of Tailwind CSS classes in structural components, since the public theme might not use Tailwind CSS.

> Also, the public UI cannot rely on any CMSUI-specific styles or components.

Reviewer: ```suggestion
Also, the public UI can't rely on any CMS UI-specific styles or components.
```
